### PR TITLE
Use FusedMatMul When Transpose is Between First Dim and Contiguous Batch Dims

### PR DESCRIPTION
**Pull request by iK1D**
Current FusedMatMul can support only Transpose on last 2 dims. When the 2-D arrays for MatMul is the 1st and last dims, and the batch dims are contiguous in the original tensor, we can also use GemmStridedBatched to calculate without doing the Transpose. The perm pattern in the Transpose is like [1,2,0,3] or [1,2,3,0]. This PR is to support these cases using FusedMatMul.

For perf comparison using a module with Add+EinSum("ks,ksm->sm")+MSELOss, K = 16, S = 7840, M = 2048, before the changes, it's ~7ms for each step, after the changes, it's ~4.5ms for each step, which has similar perf as PyTorch.

Using ULR-XL (16 layers) for perf test, before the changes, the execution graph has 195 Transpose nodes, 16 MatMul nodes and 306 FusedMatMul nodes. After the changes the numbers are: 131 Transpose nodes and 322 FusedMatMul nodes. From nvvp profiling, for each step, the execution time reduces from ~913ms to ~882ms, which have ~4%. The gain is from the reduce of Transpose compute, and the new fused FusedMatMul nodes use GemmStridedBatched, which has comparible perf as original MatMul node.s

### CODE REVIEW

**Comment by pengwa** [comment]
> Current FusedMatMul can support only Transpose on last 2 dims. When the 2-D arrays for MatMul is the 1st and last dims, and the batch dims are contiguous in the original tensor, we can also use GemmStridedBatched to calculate without doing the Transpose. The perm pattern in the Transpose is like [1,2,0,3] or [1,2,3,0]. This PR is to support these cases using FusedMatMul.
> 
> For perf comparison using a module with Add+EinSum("ks,ksm->sm")+MSELOss, K = 16, S = 7840, M = 2048, before the changes, it's ~7ms for each step, after the changes, it's ~4.5ms for each step, which has similar perf as PyTorch.
> 
> Using ULR-XL (16 layers) for perf test, before the changes, the execution graph has 195 Transpose nodes, 16 MatMul nodes and 306 FusedMatMul nodes. After the changes the numbers are: 131 Transpose nodes and 322 FusedMatMul nodes. From nvvp profiling, for each step, the execution time reduces from ~913ms to ~882ms, which have ~4%. The gain is from the reduce of Transpose compute, and the new fused FusedMatMul nodes use GemmStridedBatched, which has comparible perf as original MatMul node.s

This is a nice change!! Some site notes put here FYI. It is found APEX and other libs I investigated last week also do this trick, the trick is applied to the models having self-attention's input having shape [seq, batch, num_head, head_dim]. We would remove at least two transposes + a scaling multiple (sqrt(num_head)) for the BERT large case.   @iK1D @SherlockNoMad

File: onnxruntime/core/optimizer/matmul_transpose_fusion.cc [hunk 1]
```diff
@@ -32,14 +32,19 @@ static bool GetTransposePerms(const Node& transpose_node, std::vector<int64_t>&
   return true;
 }
 
-static Node* GetTransposeNodeFromOutput(Graph& graph, NodeArg& node_arg) {
+// is_trans is whether to transpose the 2 dims used to MatMul.
+// is_trans_batch is whether to transpose 1st dim and batch dims (dim-1 to dim-rank-2).
+// For example:
```
Comment (pengwa) [inline onnxruntime/core/optimizer/matmul_transpose_fusion.cc:37]: it would be nice if we can give a more descriptive comments covering what exact cases we target to fuse. 

 An example FYI [https://github.com/microsoft/onnxruntime/blob/91096781c34a5278c496bbd94fd0da4445a3e005/onnxruntime/core/optimizer/bias_softmax_fusion.cc#L85](https://github.com/microsoft/onnxruntime/blob/91096781c34a5278c496bbd94fd0da4445a3e005/onnxruntime/core/optimizer/bias_softmax_fusion.cc#L85)

File: onnxruntime/core/providers/rocm/math/matmul.cc [hunk 2]
```diff
@@ -110,9 +118,9 @@ Status MatMul<T>::ComputeInternal(OpKernelContext* ctx) const {
...
   rocblas_operation transB = transb ? rocblas_operation_transpose : rocblas_operation_none;
-  const int lda = transa ? static_cast<int>(helper.M()) : static_cast<int>(helper.K());
-  const int ldb = transb ? static_cast<int>(helper.K()) : static_cast<int>(helper.N());
-  const int ldc = static_cast<int>(helper.N());
+  const int lda = helper.Lda(transa);
+  const int ldb = helper.Ldb(transb);
+  const int ldc = helper.Ldc();
   int64_t stride_A, stride_B, stride_C, batch_count;
 
   if (helper.OutputOffsets().size() == 1) {
```
Comment (pengwa) [inline onnxruntime/core/providers/rocm/math/matmul.cc:126]: does this branch need "helper.FillOffsets();"?

File: onnxruntime/core/optimizer/matmul_transpose_fusion.cc (hunk 1 above)
Comment (satyajandhyala) [reply onnxruntime/core/optimizer/matmul_transpose_fusion.cc:37]: I think it is better to use a different word than "batch" because it is used with respect to training batch. May be something like "range" may be okay.

Comment (satyajandhyala) [reply onnxruntime/core/optimizer/matmul_transpose_fusion.cc:37]: Mathematically speaking this is a cycle of a fixed length less than or equal to the dimensions. (1->0, 2->1,.., r->r-1, 0->r). A transpose is a cycle of length 2.

File: onnxruntime/core/optimizer/matmul_transpose_fusion.cc [hunk 3]
```diff
@@ -48,28 +53,34 @@ static Node* GetTransposeNodeFromOutput(Graph& graph, NodeArg& node_arg) {
...
+  // We can fuse the Transpose node only when the last axis of original tensor is within the last two dims after transpose.
+  int64_t last_axis = static_cast<int64_t>(rank) - 1;
+  size_t last_axis_index = perms[rank - 1] == last_axis ? rank - 1 : perms[rank - 2] == last_axis ? rank - 2 : rank;
+  if (last_axis_index == rank) {
+    return nullptr;
   }
 
-  if (!is_trans_on_last_two_dims) {
-    return nullptr;
+  // Transpose node can be fused to MatMul when the batch dimensions have same order before and after transpose.
```
[the remark this answers is not in the document]
Comment (satyajandhyala) [reply onnxruntime/core/optimizer/matmul_transpose_fusion.cc:68]: Introducing the notion of "cycle" is really helpful to understand the code here.

File: onnxruntime/core/optimizer/matmul_transpose_fusion.cc (hunk 1 above)
Comment (satyajandhyala) [reply onnxruntime/core/optimizer/matmul_transpose_fusion.cc:37]: may be "circular permutation" is more clear.

**Comment by pengwa** [comment]
The change looks great overall! There are few things, need your help for confirmation:
1. In your measured case, how many transpose get eliminated per layer? from 195 to 131 for 16 layers, so it is 4 transpose per layer. 
2. have we ever covered the backward pass for the fused matmul?

File: onnxruntime/core/optimizer/matmul_transpose_fusion.cc (hunk 1 above)
Comment (iK1D) [reply onnxruntime/core/optimizer/matmul_transpose_fusion.cc:37]: CUDA's APIs (GemmBatched, GemmStridedBatched) use the same name. Our MatMul code also calls them batches. I think we still call batch here, but add more comments to explain.

File: onnxruntime/core/providers/cpu/math/matmul.h [hunk 4]
```diff
@@ -22,14 +22,18 @@ class MatMul<float> final : public OpKernel {
     info.GetAttrOrDefault<int64_t>("transA", &trans_a_attr_, 0);
     info.GetAttrOrDefault<int64_t>("transB", &trans_b_attr_, 0);
     info.GetAttrOrDefault<float>("alpha", &alpha_attr_, 1.0);
+    int64_t trans_batch_a_attr, trans_batch_b_attr;
+    info.GetAttrOrDefault<int64_t>("transBatchA", &trans_batch_a_attr, 0);
```
[the remark this answers is not in the document]
Comment (iK1D) [reply onnxruntime/core/providers/cpu/math/matmul.h:26]: I think boolean is better than the old style for reading. CUDA code also handle this way. If we really want to change, better change the old style.

File: onnxruntime/core/providers/cpu/math/matmul.h [hunk 5]
```diff
@@ -22,14 +22,18 @@ class MatMul<float> final : public OpKernel {
...
+    trans_batch_a_ = trans_batch_a_attr != 0;
+    trans_batch_b_ = trans_batch_b_attr != 0;
   }
 
   Status PrePack(const Tensor& tensor, int input_idx, AllocatorPtr alloc,
                  /*out*/ bool& is_packed,
                  /*out*/ PrePackedWeights* prepacked_weights) override;
 
-  Status UseSharedPrePackedBuffers(std::vector<BufferUniquePtr>& prepacked_buffers,
-                                   int input_idx,
```
[the remark this answers is not in the document]
Comment (iK1D) [reply onnxruntime/core/providers/cpu/math/matmul.h:32]: A document format caused this to follow the clang format.

File: onnxruntime/core/providers/cpu/math/matmul_helper.h [hunk 6]
```diff
@@ -42,22 +46,54 @@ class MatMulComputeHelper {
...
+        left_stride_factor_ = static_cast<size_t>(tmp);
+      }
+      if (trans_batch_b) {
+        int64_t tmp = dims_right[0];
+        for (size_t i = 0; i < right_num_dims - 2; ++i) {
+          dims_right[i] = dims_right[i + 1];
+          right_ld_factor_ *= static_cast<int>(dims_right[i]);
+        }
+        dims_right[right_num_dims - 2] = tmp;
+        right_stride_factor_ = static_cast<size_t>(tmp);
```
[the remark this answers is not in the document]
Comment (iK1D) [reply onnxruntime/core/providers/cpu/math/matmul_helper.h:90]: CUDA calls the parameter strideA/B/C directly, batch_stride is also confusing?

File: onnxruntime/core/providers/rocm/math/matmul.cc (hunk 2 above)
Comment (iK1D) [reply onnxruntime/core/providers/rocm/math/matmul.cc:126]: No need. The fillOffsets is to populate left_offsets_, right_offsets_ and output_offsets_. These are not used in this branch.

**Comment by iK1D** [comment]
> The change looks great overall! There are few things, need your help for confirmation:
> 
> 1. In your measured case, how many transpose get eliminated per layer? from 195 to 131 for 16 layers, so it is 4 transpose per layer?
> 2. have we ever covered the backward pass for the fused matmul?

I didn't check the big graph carefully, but from the number yes it's 4 for each layer. From the code the fusion is added for both training and inference transformer list, so ideally it backward is also covered. But we build the gradient graph after the training transformers, and use FusedMatMul instead of MatMul in backward graph, so I think it's rare to have such case in backward we can fuse.

**Comment by satyajandhyala** [comment]
> 
Going by the definition at https://mathworld.wolfram.com/CyclicPermutation.html
This is good. 
Going by the definition at https://mathworld.wolfram.com/CyclicPermutation.html. This is applying cyclic permutation to left by 1. Is it possible to extend the idea further, generalize for any number r less than dimensions?

**Comment by iK1D** [comment]
> > 
> 
> This is good. Going by the definition at https://mathworld.wolfram.com/CyclicPermutation.html. This is applying cyclic permutation to left by 1. Is it possible to extend the idea further, generalize for any number r less than dimensions?

I don't quite get the idea. Could you please give some example? i.e., what's the 'perm' attribute for the Transpose nodes.

**Comment by satyajandhyala** [comment]
> > > 
> > 
> > 
> > This is good. Going by the definition at https://mathworld.wolfram.com/CyclicPermutation.html. This is applying cyclic permutation to left by 1. Is it possible to extend the idea further, generalize for any number r less than dimensions?
> 
> I don't quite get the idea. Could you please give some example? i.e., what's the 'perm' attribute for the Transpose nodes.

This change supports [1,2,0,3] or [1,2,3,0]. In future, not in this PR, could consider permutations like, [2, 0, 1, 3] or [2,3,0,1] also?

**Comment by iK1D** [comment]
> > > > 
> > > 
> > > 
> > > This is good. Going by the definition at https://mathworld.wolfram.com/CyclicPermutation.html. This is applying cyclic permutation to left by 1. Is it possible to extend the idea further, generalize for any number r less than dimensions?
> > 
> > 
> > I don't quite get the idea. Could you please give some example? i.e., what's the 'perm' attribute for the Transpose nodes.
> 
> This change supports [1,2,0,3] or [1,2,3,0]. In future, not in this PR, could we consider permutations like, [2, 0, 1, 3] or [2,3,0,1] also?

I have below comments in the code to explain which cases we can fuse. For [2,0,1,3] or [2,3,0,1], it's not possible to get the strideA, strideB, lda, ldb for the parameters of GemmStridedBatched, so we cannot fuse such cases.

  // Transpose node can be fused to MatMul when the batch dims keep same relative orders before and after transpose.
  // But if they are not contiguous, after the fusion, we can only use GemmBatched instead of GemmStridedBatched,
  // which may have perf issue. To keep it simple, we will fuse only when batch dimensions are contiguous.

File: onnxruntime/core/providers/cpu/math/matmul.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (pengwa) [reply onnxruntime/core/providers/cpu/math/matmul.h:26]: Okay.

**Review by pengwa — APPROVED** [review]
Sorry for the late response! LGTM!! :)
